In [1]:
from google.colab import drive
drive.mount('/gdrive')
%cd /gdrive

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /gdrive
/gdrive


In [0]:
import pandas as pd

In [0]:
df = pd.read_csv("/content/drive/My Drive/A.I/udemy machine learning/seccion 7 regresion lineal/92 enmascarado de variables categoricas redundantes/Ecom Expense.csv")

In [5]:
df.head()

Transaction ID  Age    Items   ...  Gender  City Tier  Total Spend
0         TXN001    42       10  ...  Female     Tier 1  4198.385084
1         TXN002    24        8  ...  Female     Tier 2  4134.976648
2         TXN003    47       11  ...    Male     Tier 2  5166.614455
3         TXN004    50       11  ...  Female     Tier 1  7784.447676
4         TXN005    60        2  ...  Female     Tier 2  3254.160485

[5 rows x 9 columns]

ingresos de entrada y salida para crear un modelo para predecir los gastos deacuerdo al genero y el sector de la ciudad

convertir las variable categoricas a numericas

In [0]:
dummy_gender = pd.get_dummies(df["Gender"], prefix="Gender")
dummy_city_tier = pd.get_dummies(df["City Tier"], prefix= "City")

In [7]:
dummy_gender.head()

Gender_Female  Gender_Male
0              1            0
1              1            0
2              0            1
3              1            0
4              1            0

In [8]:
dummy_city_tier.head()

City_Tier 1  City_Tier 2  City_Tier 3
0            1            0            0
1            0            1            0
2            0            1            0
3            1            0            0
4            0            1            0

In [9]:
df.head()

Transaction ID  Age    Items   ...  Gender  City Tier  Total Spend
0         TXN001    42       10  ...  Female     Tier 1  4198.385084
1         TXN002    24        8  ...  Female     Tier 2  4134.976648
2         TXN003    47       11  ...    Male     Tier 2  5166.614455
3         TXN004    50       11  ...  Female     Tier 1  7784.447676
4         TXN005    60        2  ...  Female     Tier 2  3254.160485

[5 rows x 9 columns]

In [10]:
df.columns

Index(['Transaction ID', 'Age ', ' Items ', 'Monthly Income',
       'Transaction Time', 'Record', 'Gender', 'City Tier', 'Total Spend'],
      dtype='object')

In [11]:
type(df.columns)

pandas.core.indexes.base.Index

hacer join para unir los dataframes con el dataframe principal

pasar a lista y ver la sintaxis de los nombres de las columnas

In [12]:
column_names = df.columns.values.tolist()
column_names

['Transaction ID',
 'Age ',
 ' Items ',
 'Monthly Income',
 'Transaction Time',
 'Record',
 'Gender',
 'City Tier',
 'Total Spend']

In [13]:
type(column_names)

list

items y age tiene espacios en blanco

unir el primer dataframe

In [14]:
df_new = df[column_names].join(dummy_gender)
df_new.head()

Transaction ID  Age    Items   ...  Total Spend  Gender_Female  Gender_Male
0         TXN001    42       10  ...  4198.385084              1            0
1         TXN002    24        8  ...  4134.976648              1            0
2         TXN003    47       11  ...  5166.614455              0            1
3         TXN004    50       11  ...  7784.447676              1            0
4         TXN005    60        2  ...  3254.160485              1            0

[5 rows x 11 columns]

In [15]:
column_names

['Transaction ID',
 'Age ',
 ' Items ',
 'Monthly Income',
 'Transaction Time',
 'Record',
 'Gender',
 'City Tier',
 'Total Spend']

pasar las columnas nuevas al column names

In [16]:
column_names = df_new.columns.values.tolist()
column_names

['Transaction ID',
 'Age ',
 ' Items ',
 'Monthly Income',
 'Transaction Time',
 'Record',
 'Gender',
 'City Tier',
 'Total Spend',
 'Gender_Female',
 'Gender_Male']

unir el dataframe con el otro dummy faltante

In [17]:
df_new = df_new[column_names].join(dummy_city_tier)
column_names = df_new.columns.values.tolist()
column_names

['Transaction ID',
 'Age ',
 ' Items ',
 'Monthly Income',
 'Transaction Time',
 'Record',
 'Gender',
 'City Tier',
 'Total Spend',
 'Gender_Female',
 'Gender_Male',
 'City_Tier 1',
 'City_Tier 2',
 'City_Tier 3']

In [18]:
df_new.head()

Transaction ID  Age    Items   ...  City_Tier 1  City_Tier 2  City_Tier 3
0         TXN001    42       10  ...            1            0            0
1         TXN002    24        8  ...            0            1            0
2         TXN003    47       11  ...            0            1            0
3         TXN004    50       11  ...            1            0            0
4         TXN005    60        2  ...            0            1            0

[5 rows x 14 columns]

variables del predecitoras

In [0]:
feature_cols = ["Monthly Income", "Transaction Time",
                "Gender_Female", "Gender_Male", "City_Tier 1",
                "City_Tier 2", "City_Tier 3", "Record"]

variable a predecir en y

In [0]:
X = df_new[feature_cols]
y = df_new["Total Spend"]

crear el modelo con sklearn

In [0]:
from sklearn.linear_model import LinearRegression

In [0]:
lm = LinearRegression()

In [23]:
lm.fit(X, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [24]:
print(lm.intercept_)

-79.41713030137544


In [25]:
print(lm.coef_)

[ 1.47538980e-01  1.54946125e-01 -1.31025013e+02  1.31025013e+02
  7.67643260e+01  5.51389743e+01 -1.31903300e+02  7.72233446e+02]


juntar los valores de los coeficientes en uno solo

In [26]:
zip(feature_cols, lm.coef_)

In [27]:
list(zip(feature_cols, lm.coef_))

[('Monthly Income', 0.14753898049205746),
 ('Transaction Time', 0.15494612549589704),
 ('Gender_Female', -131.02501325554584),
 ('Gender_Male', 131.02501325554567),
 ('City_Tier 1', 76.76432601049501),
 ('City_Tier 2', 55.13897430923278),
 ('City_Tier 3', -131.90330031972783),
 ('Record', 772.233445744565)]

que tan bueno es el modelo

In [28]:
lm.score(X, y)

0.9179923586131016

el modelo puede ser escrito como:

y(toal spend) = -79.41713030137544 + 

Monthly Income * 0.14753898049205746 + 

Transaction Time * 0.15494612549589704 + 

Gender_Female * -131.02501325554584 +

Gender_Male * 131.02501325554567 +

City_Tier 1 * 76.76432601049501 +

City_Tier 2 * 55.13897430923278 +

City_Tier 3 * -131.90330031972783 +
 
 Record * 772.233445744565

hacer las predicciones

In [0]:
predictions = lm.predict(X)

In [0]:
df_new["predictions"] = predictions

In [31]:
df_new.head()

Transaction ID  Age    Items   ...  City_Tier 2  City_Tier 3  predictions
0         TXN001    42       10  ...            0            0  4903.696720
1         TXN002    24        8  ...            1            0  4799.434826
2         TXN003    47       11  ...            1            0  5157.082504
3         TXN004    50       11  ...            0            0  8068.012996
4         TXN005    60        2  ...            1            0  3581.980335

[5 rows x 15 columns]

suma de los caudrados de las diferencias

In [0]:
import numpy as np

suma de los cuadrados de las diferencias

In [0]:
SSD = np.sum((df_new["predictions"] - df_new["Total Spend"])**2)

In [34]:
SSD

1517733985.3408163

Desviacion tipica de los residuos

In [0]:
RSE = np.sqrt(SSD / (len(df_new) - len(feature_cols)-1))

In [36]:
RSE

803.1318809818165

Desviacion de 803 pesos por arriba y 803 pesos por abajo

Promedio de las ventas

In [0]:
sales_mean = np.mean(df_new["Total Spend"])

In [38]:
sales_mean

6163.176415976714

media del error

In [0]:
error = RSE / sales_mean * 100

In [40]:
error

13.031135680294161

**Eliminar variables dummy redundantes, la variable categorica que queda por defecto**

el cambio repercutira en las alfas, intercepto

In [41]:
dummy_gender = pd.get_dummies(df["Gender"], prefix="Gender").iloc[:,1:]
dummy_gender.head()

Gender_Male
0            0
1            0
2            1
3            0
4            0

In [42]:
dummy_city_tier = pd.get_dummies(df["City Tier"], prefix="City").iloc[:,1:]
dummy_city_tier.head()

City_Tier 2  City_Tier 3
0            0            0
1            1            0
2            1            0
3            0            0
4            1            0

In [43]:
column_names = df.columns.values.tolist()
df_new = df[column_names].join(dummy_gender)
column_names = df_new.columns.values.tolist()
df_new = df_new[column_names].join(dummy_city_tier)
df_new.head()

Transaction ID  Age    Items   ...  Gender_Male  City_Tier 2  City_Tier 3
0         TXN001    42       10  ...            0            0            0
1         TXN002    24        8  ...            0            1            0
2         TXN003    47       11  ...            1            1            0
3         TXN004    50       11  ...            0            0            0
4         TXN005    60        2  ...            0            1            0

[5 rows x 12 columns]

**seleccionar columnas para el modelo**

In [0]:
feature_cols = ["Monthly Income", "Transaction Time", "Gender_Male", "City_Tier 2", "City_Tier 3", "Record"]

In [0]:
X = df_new[feature_cols]

**Selecciono variable a predecir "Total Spend"**

In [0]:
y = df_new["Total Spend"]

**creo el modelo de regresion lineal**

In [0]:
lm = LinearRegression()

In [50]:
lm.fit(X,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

**ver el intercepto**

In [51]:
lm.intercept_

-133.67781754642238

**listar el modelo con los coeficientes**

In [52]:
list(zip(feature_cols, lm.coef_))

[('Monthly Income', 0.1475389804920575),
 ('Transaction Time', 0.1549461254958966),
 ('Gender_Male', 262.0500265110957),
 ('City_Tier 2', -21.625351701263007),
 ('City_Tier 3', -208.6676263302229),
 ('Record', 772.2334457445636)]

**desempeño del modelo**

In [54]:
lm.score(X, y)

0.9179923586131016